clang/lib/Sema/SemaExpr.cpp

```diff
@@ -19549,7 +19549,17 @@ static bool captureInLambda(LambdaScopeInfo *LSI, ValueDecl *Var,
   }
 
   // Compute the type of the field that will capture this variable.
-  if (ByRef) {
+  if (LSI->ImpCaptureStyle == LambdaScopeInfo::ImpCap_LambdaByCallable) {
+     Expr *E = DeclRefExpr::Create(S.Context, NestedNameSpecifierLoc(), Loc,
+                                   Var, false, Loc, CaptureType, VK_LValue);
+     MultiExprArg arg(&E, 1);
+     ExprResult CaptureInit = S.ActOnCallExpr(S.getCurScope(),
+                                              LSI->DefaultCallable.get(),
+                                              Loc, arg, Loc);
+     CaptureType = CaptureInit.get()->getType();
+     ByRef = CaptureType->isReferenceType();
```

Comment:
isLValueReferenceType()?

Reply:
If it is not by reference, I think we still want this check:
// Make sure that by-copy captures are of a complete and non-abstract type.

Reply:
If the function returns an rvalue, then I think that should be treated as capture by copy.